### PR TITLE
Add support for twice-differentiable DTensor redistribution

### PR DESCRIPTION
**Pull request by jvwilliams23**
Fixes #160313

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @xmfan @H-Huang

### CODE REVIEW

**Review by sladyn98 — APPROVED** [review]
Consider updating the docstring for the `redistribute()` method in `_api.py:537` to mention that it now supports second-order derivatives:

**Comment by jvwilliams23** [comment]
@sladyn98 thanks for the review. Conflicts resolved and changes incorporated.

**Comment by norabelrose** [comment]
This feature would also be useful for a research project I am currently working on, so it would be great if this can be merged into main soon.

File: torch/distributed/tensor/_redistribute.py [hunk 1]
```diff
@@ -947,15 +977,76 @@ def backward(ctx, grad_output: "dtensor.DTensor"):  # type: ignore[override]
...
+
+        return dtensor.DTensor(
             output,
             spec,
             # pyrefly: ignore [unexpected-keyword]
             requires_grad=grad_output.requires_grad,
         )
 
+    @staticmethod
+    def backward(ctx, grad2_output: "dtensor.DTensor"):  # type: ignore[override]
```
Comment (vishal9-team) [inline torch/distributed/tensor/_redistribute.py:1884]: Any reason why we're replicating backward here - could it not be refactored into common redistribute_backward?

Comment (jvwilliams23) [reply torch/distributed/tensor/_redistribute.py:1884]: Good point, this is fixed now.

File: torch/distributed/tensor/_api.py [hunk 2]
```diff
@@ -116,12 +116,11 @@ def backward(ctx, grad_output: torch.Tensor):  # type: ignore[override]
             )
         return (
             # pyrefly: ignore [bad-argument-type]
-            DTensor(
+            DTensor.from_local(
```
Comment (vishal9-team) [inline torch/distributed/tensor/_api.py:153]: @zpcore does the StridedShard cases? cc: @wconstab

File: torch/distributed/tensor/_redistribute.py [hunk 3]
```diff
@@ -1021,76 +1108,71 @@ def forward(  # type: ignore[override]
...
-            if (current.is_shard() or current.is_replicate()) and target.is_partial():
-                normalized_placements.append(Replicate())
-            else:
-                normalized_placements.append(target)
 
-        previous_spec = DTensorSpec(
-            previous_spec.device_mesh,
-            placements=tuple(normalized_placements),
-            tensor_meta=previous_spec.tensor_meta,
+class RedistributeBackward(torch.autograd.Function):
```
Comment (vishal9-team) [inline torch/distributed/tensor/_redistribute.py]: 1. Can you call it NestedRedistribute?
2. Can you have NestedRedistribute::backward call NestedRedistribute::apply?

@albanD - do you see any issue with that

Comment (jvwilliams23) [reply torch/distributed/tensor/_redistribute.py]: Done

File: torch/distributed/tensor/_api.py (hunk 2 above)
Comment (wconstab) [reply torch/distributed/tensor/_api.py:153]: didn't understand this comment

File: torch/distributed/tensor/_redistribute.py [hunk 4]
```diff
@@ -1021,78 +1108,68 @@ def forward(  # type: ignore[override]
...
+        ctx.async_op = async_op
+        ctx.backward_dtype = backward_dtype or ctx.original_dtype
+        ctx.original_dtype = grad_output._local_tensor.dtype
 
-        output = redistribute_local_tensor(
-            local_tensor,
-            current_spec,
-            previous_spec,
-            async_op=async_op,
+        output, spec = _redistribute_backward(
```
Comment (wconstab) [inline torch/distributed/tensor/_redistribute.py:1868]: is this actually the first backward pass and then the 'backward' of this autograd function is the second reentrant backward pass? or why is this called 'backward' and happening during the .forward method? a docstring explaining the usage of NestedRedistribute would be good.

File: torch/distributed/tensor/_api.py (hunk 2 above)
Comment (vishal9-team) [reply torch/distributed/tensor/_api.py:153]: To clarify my earlier comment - the main logic difference between calling DTensor() directly vs DTensor.from_local() is  ostly in the handling of strided/shard placements (specifically StridedShard normalization and validation). I want to make sure this change is effectively a no-op for the redistribution backward case.

Since the backward pass here is working with grad tensors that should already have normalized placements from the forward, I believe this should be safe - but wanted to confirm we're not accidentally triggering extra validation or placement transformations that could affect correctness or perf.

@zpcore  - can you confirm from_local is semantically equivalent here given the context?

Comment (wconstab) [reply torch/distributed/tensor/_api.py:153]: ah, that's an interesting comment, I didn't consider it before. I am not sure why we want to have a semantic difference relating to stridedshard between DTensor() and from_local(). is it intentional?

Comment (zpcore) [reply torch/distributed/tensor/_api.py:153]: I think Shard/_StridedShard should already be normalized when they reach _ToTorchTensor.forward. It seems OK to me using `DTensor.from_local()`.

But what's the benefit of using `DTensor.from_local` here?

Comment (zpcore) [reply torch/distributed/tensor/_api.py:153]: I see, they need the second order gradient and we need `from_local` to provide the backward function.

File: torch/distributed/tensor/_redistribute.py [hunk 5]
```diff
@@ -1021,78 +1108,68 @@ def forward(  # type: ignore[override]
...
             requires_grad=grad_output.requires_grad,
         )
 
+    @staticmethod
+    def backward(ctx, grad2_output: "dtensor.DTensor"):  # type: ignore[override]
+        previous_spec = ctx.current_spec
+        async_op = ctx.async_op
+        backward_dtype = ctx.backward_dtype or ctx.original_dtype
+
+        output_dtensor = NestedRedistribute.apply(
```
Comment (zpcore) [inline torch/distributed/tensor/_redistribute.py:1889]: TIL the autograd function can apply to itself :)

File: torch/distributed/tensor/_redistribute.py (hunk 1 above)
Comment (zpcore) [reply torch/distributed/tensor/_redistribute.py:1884]: I only saw `_redistribute_backward` been called once.

File: torch/distributed/tensor/_redistribute.py (hunk 4 above)
Comment (jvwilliams23) [reply torch/distributed/tensor/_redistribute.py:1868]: Yes this is the first backward pass, then the `backward` call will do the second backward pass (to get second-order gradient).
Sure, I will add a docstring no problem.

File: torch/distributed/tensor/_redistribute.py (hunk 1 above)
Comment (jvwilliams23) [reply torch/distributed/tensor/_redistribute.py:1884]: It was called twice until we added `NestedRedistribute` [cb7e49c](https://github.com/pytorch/pytorch/pull/160509/commits/cb7e49cdb9335c5e3d3afc47f4e976e49fedfad9). 

I can move the code from `_redistribute_backward` to `NestedRedistribute.forward`?

File: torch/distributed/tensor/_redistribute.py [hunk 6]
```diff
@@ -1021,76 +1108,71 @@ def forward(  # type: ignore[override]
...
+            output,
+            spec,
+            # pyrefly: ignore [unexpected-keyword]
+            requires_grad=grad_output.requires_grad,
         )
 
-        if output.dtype != ctx.original_dtype:
-            output = output.to(ctx.original_dtype)
+    @staticmethod
+    def backward(ctx, grad2_output: "dtensor.DTensor"):  # type: ignore[override]
```
Comment (albanD) [inline torch/distributed/tensor/_redistribute.py:1884]: This backward all is not differentiable itself. So we should mark it as such (triple backward won't work).

File: torch/distributed/tensor/_redistribute.py (hunk 3 above)
Comment (albanD) [reply torch/distributed/tensor/_redistribute.py]: Given that this doesn't get you to be infinitely differentiable, I guess RedistributeBackward is ok as a name.

You can indeed make the double backward differentiable by making this Function's backward differentiable. In such a case, the rename would make sense indeed.

File: torch/distributed/tensor/_redistribute.py (hunk 6 above)
Comment (jvwilliams23) [reply torch/distributed/tensor/_redistribute.py:1884]: I have added a docstring for NestedRedistribute with a note that triple backward won't work.

File: torch/distributed/tensor/_redistribute.py (hunk 4 above)
Comment (jvwilliams23) [reply torch/distributed/tensor/_redistribute.py:1868]: Done @wconstab

File: torch/distributed/tensor/_redistribute.py (hunk 3 above)
Comment (jvwilliams23) [reply torch/distributed/tensor/_redistribute.py]: Do you guys want me to change the name back to `RedistributeBackward` or leave as `NestedRedistribute`?

**Comment by norabelrose** [comment]
I'd like to bump this PR; I can vouch that it is working for my needs.

**Review by wconstab — COMMENTED** [review]
lgtm, thanks for the fix! you'll probably need to rebase this

File: torch/distributed/tensor/_redistribute.py [hunk 7]
```diff
@@ -1674,6 +1674,93 @@ def redistribute_local_tensor(
...
+
+    # for backward shard -> partial, we just do shard -> replicate
+    # for backward replicate -> partial, we skip the transformation
+    normalized_placements: list[Placement] = []
+    for current, target in zip(current_spec.placements, previous_spec.placements):
+        if (current.is_shard() or current.is_replicate()) and target.is_partial():
+            normalized_placements.append(Replicate())
+        else:
+            normalized_placements.append(target)
+
```
Comment (luciaquirke) [inline torch/distributed/tensor/_redistribute.py:1735]: @jvwilliams23 Hello, thanks for this feature, we've been getting some great use out of it :) 

I think we may need to apply the local gradient of `1/world_size` here when `reduce_op="avg"`:

```suggestion
    avg_dims: list[int] = []
    for i, (current, target) in enumerate(zip(current_spec.placements, previous_spec.placements)):
        if (current.is_shard() or current.is_replicate()) and target.is_partial():
            normalized_placements.append(Replicate())
            if cast(Partial, target).reduce_op == "avg":
                avg_dims.append(i)
        else:
            normalized_placements.append(target)
            
    for dim in avg_dims:
        local_tensor = local_tensor / previous_spec.device_mesh.size(dim)
```

Comment (jvwilliams23) [reply torch/distributed/tensor/_redistribute.py:1735]: Thanks for the review @luciaquirke, glad you are finding it useful! Could you please add a test for that if you have time?
